src/elastio-snap.c

```diff
@@ -5186,6 +5185,93 @@ static int __handle_bdev_mount_writable(const char __user *dir_name, const struc
 	return ret;
 }
 
+// The function below is called to gain or restore the ownership over a block device and resolves
+// the following bug related to the dormant snapshot functionality:
+//
+// [ Issue https://github.com/elastio/elastio-snap/issues/126 ]
+//
+//  - at the first mount, the block device is owned by the loop driver
+//  - after the snapshot creation, we update ownership to elastio driver
+//  - because of this, v5.13+ kernels did additional module_put() for our module at umount
+//
+//  To resolve this, when active and umount, we switch the ownership back to the parent, hence
+//  making the umount independent from the perspective of the elastio driver. When the device
+//  is mounted again, re-gain ownership over it.
+//
+//  Apart from it, we do this to prevent the kernel from freeing our block_device_operations
+//  installed (f.e, if the device is removed from the system)
+
+#define OWNERSHIP_TO_PARENT 0
+#define OWNERSHIP_TO_DRIVER 1
+
+static int bdev_switch_ownership(const char __user *dir_name, int follow_flags, int ownership)
+{
+	int i, ret = 0;
+	int lookup_flags = 0;
+	struct path path = {};
+	struct snap_device *dev;
+	struct block_device *bdev;
+
+	if(!(follow_flags & UMOUNT_NOFOLLOW)) lookup_flags |= LOOKUP_FOLLOW;
+
+	ret = user_path_at(AT_FDCWD, dir_name, lookup_flags, &path);
+	if(ret){
+		//error finding path
+		goto out;
+	}else if(path.dentry != path.mnt->mnt_root){
+		//path specified isn't a mount point
+		ret = -ENODEV;
+		goto out;
+	}
+
+	bdev = path.mnt->mnt_sb->s_bdev;
+	if(!bdev){
+		//path specified isn't mounted on a block device
+		ret = -ENODEV;
+		goto out;
+	}
+
+	tracer_for_each(dev, i) {
+		if (!dev) continue;
+		if (dev->sd_base_dev == bdev) break;
+	}
+
+	// not found
+	if (!dev) {
+		LOG_DEBUG("no active snap device found, skip");
+		ret = 0;
+		goto out;
+	}
+
+	switch (ownership) {
+		// umount
+		case OWNERSHIP_TO_PARENT:
+			LOG_DEBUG("Restoring bdev ownership");
```

Comment:
It would be great to see here bdev name or a mount point. The second you have already. The first you can get like this
https://github.com/elastio/elastio-snap/blob/c197dab8cb87d261969dcc9dcba478d434b633cf/src/elastio-snap.c#L3852-L3854